Added tests for AverageModelCheckpoint() callback

ashutosh1919 opened the pull request:
Fixes #1502 .  
@gabrieldemarmiesse, @shun-lin and @Squadrick , Please review.

ashutosh1919 commented:
@shun-lin , @Squadrick - The build is cancelled. I don't understand. Please help

gabrieldemarmiesse commented:
That behavior from github actions is very unexpected. Either the tensorflow org has run out of build minutes, either there is a problem with github itself. I'll try to restart the CI.

AakashKumarNain commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+try:
+    import h5py
+except ImportError:
+    h5py = None
```
Please add a message in the exception

gabrieldemarmiesse replied:
Actually I believe we can remove this import as it's not used in the file.

AakashKumarNain commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+from tensorflow import keras
+from tensorflow_addons.callbacks import AverageModelCheckpoint
+from tensorflow_addons.optimizers import MovingAverage
+import os
```
`os` should be imported at the top. Please refer to [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) for more information

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+EPOCHS = 5
+test_model_filepath = "test_model.h5"
+
+moving_avg = MovingAverage(
+    tf.keras.optimizers.SGD(lr=2.0), sequential_update=True, average_decay=0.5,
+)
```
Please move the optimizer in the `get_data_and_model`, tests should not interfer with each other, and as such, they should not modify any global object (in this case an optimizer). The optimizer should be re-created in each test.

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+    with pytest.raises(
+        TypeError,
+        match="AverageModelCheckpoint is only used when trainingwith"
+        + " MovingAverage or StochasticAverage",
```
```suggestion
        " MovingAverage or StochasticAverage",
```

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+        match="AverageModelCheckpoint is only used when trainingwith"
+        + " MovingAverage or StochasticAverage",
+    ):
+        assert model.fit(
```
```suggestion
        model.fit(
```

That would try to convert the keras history into a boolean object. Tests don't necesarily need assert statements.

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+
+def test_save_best_only():
+    x, y, model = get_data_and_model()
+    save_best_only = True
```
Here we test that the callback runs, which is good, but we could also check for correctness (the callback does what we want it to do). 

To do that, grab the keras history in the fit, load the saved model, do an evaluation pass and ensure the metric is equal to the best one found in the keras history.


I don't know if it's possible to do that for other tests, but if it's possible, we should aim for that.

ashutosh1919 replied:
@gabrieldemarmiesse , please refer to [this comment](https://github.com/keras-team/keras/issues/6977#issuecomment-400207482). I don't think it is a good idea to test `fit()` results with `evaluate()`

gabrieldemarmiesse replied:
You're right, you'll have problems if you compare on the train data. On the validation data, the keras history should give you the same result as a standalone evaluate call since the model does't change during an evaluation pass.

ashutosh1919 replied:
FYI, all types of data here are same. And I have tested it and after testing it, I can't strongly suggest that method because the results I was getting are not good.

ashutosh1919 replied:
Maybe We should use any particular simple data which doesn't deviate the model.

AakashKumarNain replied:
@ashutosh1919 can you paste your code here which you have tested?

gabrieldemarmiesse replied:
If it doesn't work, it may be that there is a bug in the implementation of the callback, in which case we should decorate the test with `@pytest.xfail` and open an issue to track the bug. @ashutosh1919 we won't ask you to fix the bug, it's outside the scope of this pull request.

ashutosh1919 replied:
@gabrieldemarmiesse , I think I got the idea of testing this. We need to test the model weights. Got the idea from some tests in tf repo.

AakashKumarNain replied:
I would have expected that weird behavior in earlier versions because of how batchnorm used to operate but that has been resolved long ago. So, IMO, `evaluate` would give you the same result as during the `fit`

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+NUM_HIDDEN = 5
+BATCH_SIZE = 5
+EPOCHS = 5
+test_model_filepath = "test_model.h5"
```
Using the same path for all the tests won't work if the tests are all run concurrently (with pytest-xdist for example). We also don't want to pollute the current working directory if one test fails and the file is not removed. The clean solution is to use this fixture: https://docs.pytest.org/en/latest/tmpdir.html#the-tmp-path-fixture , we also won't have to do `os.remove`.

ashutosh1919 replied:
done

shun-lin commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,206 @@
+    x = tf.random.normal([TRAIN_SAMPLES, INPUT_DIM])
+    y = tf.random.normal([TRAIN_SAMPLES, NUM_CLASSES])
+    inputs = keras.layers.Input(INPUT_DIM)
+    layer1 = keras.layers.Dense(NUM_HIDDEN, input_dim=INPUT_DIM, activation="relu")(
```
NIT: since you only have 1 layer maybe this variable to `hidden_layer` is better for readability.

ashutosh1919 commented:
@gabrieldemarmiesse , @AakashKumarNain , @shun-lin - Resolved all the reviews except one and modified the code accordingly. The review which is puzzling is to test the functionality of the model. Please give some ideas. I will also try to experiment.

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,199 @@
+    return x, y, model
+
+
+def test_compatibility_with_some_opts_only(tmpdir):
```
Please use `tmp_path` instead of `tmpdir`, the reason is that it uses pathlib, which is a modern library to handle paths. Going forward, we're going to try to remove progressively the usage of `os.path` in addons to make the codebase more readable and bug-free.

AakashKumarNain replied:
Yes, we should rely on `pathlib`

ashutosh1919 commented:
@gabrieldemarmiesse , Changed `tmpdir` to `tmp_path`. Also, added last function to test functionality of model by asserting model weights. Please review now.

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,263 @@
+    return model, train_ds, filepath, weights_after_one_more_epoch
+
+
+def get_checkpoint_load_weights_on_restart_true_test():
```
test functions should have this form: `test_*`. Otherwise they're not run.

ashutosh1919 replied:
Sorry, didn't notice while naming.

gabrieldemarmiesse commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,263 @@
+    np.testing.assert_almost_equal(
+        weights_after_one_more_epoch, weights_after_model_restoring_and_one_more_epoch
+    )
+    not np.testing.assert_almost_equal(
```
```suggestion
    np.testing.assert_almost_equal(
```

The not has no effect here. Overall those tests show that the model doesn't change during the last epochs, I'm not sure this is what the callback does. I may be mistaken.

ashutosh1919 replied:
not is put because the weights will be different both times. Let me see another fn.

gabrieldemarmiesse commented:
@googlebot I consent.

ashutosh1919 commented on tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py:
```diff
@@ -0,0 +1,267 @@
+    return model, train_ds, filepath, weights_after_one_more_epoch
+
+
+def test_checkpoint_load_weights(tmp_path):
```
@gabrieldemarmiesse , You are right. We need to think in a better direction.
